common/metrics/grpc.go

```diff
@@ -0,0 +1,156 @@
+// metrics propagation context.
+func NewServerMetricsContextInjectorInterceptor() grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
+		ctxWithMetricsBaggage := AddMetricsBaggageToContext(ctx)
+		return handler(ctxWithMetricsBaggage, req)
+	}
+}
+
+// NewClientMetricsTrailerPropagatorInterceptor returns grpc client interceptor that injects metrics propagation context
+// received in trailer into golang metrics propagation context.
+func NewClientMetricsTrailerPropagatorInterceptor(logger log.Logger) grpc.UnaryClientInterceptor {
+	return func(
+		ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker,
+		opts ...grpc.CallOption,
+	) error {
+		var trailer metadata.MD
+		optsWithTrailer := append(opts, grpc.Trailer(&trailer))
+		err := invoker(ctx, method, req, reply, cc, optsWithTrailer...)
+
+		metricsBaggageStrings := trailer.Get(baggageTrailerKey)
+		if metricsBaggageStrings == nil {
+			return err
+		}
+
+		for _, str := range metricsBaggageStrings {
+			data := []byte(str)
+			metricsBaggage := &metricspb.Baggage{}
+			unmarshalErr := metricsBaggage.Unmarshal(data)
+			if unmarshalErr != nil {
+				logger.Error("unable to unmarshal metrics baggage from trailer", tag.Error(unmarshalErr))
+				continue
+			}
+			for key, value := range metricsBaggage.CountersInt {
+				ContextCounterAdd(ctx, logger, key, value)
+			}
+		}
+
+		return err
+	}
+}
+
+// NewServerMetricsTrailerPropagatorInterceptor returns grpc server interceptor that injects metrics propagation context
+// into gRPC trailer.
+func NewServerMetricsTrailerPropagatorInterceptor(logger log.Logger) grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
+		// we want to return original handler response, so don't override err
+		resp, err := handler(ctx, req)
+
+		baggage := GetMetricsBaggageFromContext(ctx)
+		if baggage == nil {
+			return resp, err
+		}
+
+		bytes, marshalErr := baggage.Marshal()
+		if marshalErr != nil {
+			logger.Error("unable to marshal metric baggage", tag.Error(marshalErr))
+		}
+
+		md := metadata.Pairs(baggageTrailerKey, string(bytes))
+		marshalErr = grpc.SetTrailer(ctx, md)
+		if marshalErr != nil {
+			logger.Error("unable to add metrics baggage to gRPC trailer", tag.Error(marshalErr))
+		}
+
+		return resp, err
+	}
+}
+
+// GetMetricsBaggageFromContext extracts propagation context from go context.
+func GetMetricsBaggageFromContext(ctx context.Context) *metricspb.Baggage {
+	metricValues := ctx.Value(metricsCtxKey)
+	if metricValues == nil {
+		return nil
+	}
+
+	metricValuesCtx, ok := metricValues.(*metricspb.Baggage)
+	if !ok {
+		return nil
+	}
+	return metricValuesCtx
```

Comment:
```suggestion
	return metricValues.(*metricspb.Baggage)
```
this will panic if `metricValues` are not of type `*metricspb.Baggage` which should never happen.